Added xep_radar_connector.py

beduffy opened the pull request:
- class xep_radar_connector with constructor params for iterations, pps, area start and end, port, dac_min and max
- some serial timeout issues highlighted
- hardcoding of prf_div to 25
- slightly variable calculation of the number of range bins with fasttime_ax variable but slmx4.numSamplers is always correct
- most functions unfilled from copying over from matlab
- important functions translated from matlab: StartServer, updateNumberOfSamplers, ListVariables, GetFrameNormalizedDouble, GetFrameNormalized, GetFrameRaw, Item, TryUpdateChip, Close, Open, sys_init, __ init __
- new useful functions for get_data_matrix, get_apdata

Forgive us for some of the code for being messy, but it has been working well for us for our use case. The main issues I suppose are:
- unfilled functions that we aren't using
- timeout constantly happens 15000 times while reading frames. Copied this over from matlab but matlab might deal with timeouts differently. For months we have had consistent FPS but recently not so we're looking into if this is the cause (most likely something else). Either way it should be easy to fix and an empty loop of 15000 iterations is very fast in Python.
- general comments and messiness. 

Let me know if you would like any more details or have any questions. Happy to make some improvements e.g. remove or add anything.

beduffy committed:
Added xep_radar_connector.py

- class xep_radar_connector with constructor params for iterations, pps, area start and end, port, dac_min and max
- some serial timeout issues highlighted
- hardcoding of prf_div to 25
- slightly variable calculation of the number of range bins with fasttime_ax but slmx4.numSamplers is ground truth
- most functions unfilled from copying over from matlab
- important functions duplicated: StartServer, updateNumberOfSamplers, ListVariables, GetFrameNormalizedDouble, GetFrameNormalized, GetFrameRaw, Item, TryUpdateChip, Close, Open, sys_init, __init__
- new useful functions for get_data_matrix, get_apdata

flatearth-hadella commented:
This is pretty cool, but had trouble getting it to work today.

Please add directions at the top to install the three dependencies:
```
pip install numpy
pip install matplotlib
pip install pyserial
```

This was the output I saw:
```
$ python3 xep_radar_connector.py 
starting server
ser open: False
Status after NVA_CreateHandle() b'<ACK>'
Status after OpenRadar('X4'): b''
self.x4DownConverter 1
b'VarSetValue_ByName(ddc_en, 1)'
Status after TryUpdateChip(): b'<ACK><ACK>'
Traceback (most recent call last):
  File "xep_radar_connector.py", line 389, in <module>
    slmx4.Open('X4')
  File "xep_radar_connector.py", line 112, in Open
    self.sys_init()
  File "xep_radar_connector.py", line 42, in sys_init
    self.TryUpdateChip('ddc_en', 1)
  File "xep_radar_connector.py", line 137, in TryUpdateChip
    self.updateNumberOfSamplers()
  File "xep_radar_connector.py", line 298, in updateNumberOfSamplers
    self.numSamplers = self.Item('SamplersPerFrame')
  File "xep_radar_connector.py", line 166, in Item
    angle_bracket_idx = chr_arr.index('<') if '<' in chr_arr else -1
TypeError: a bytes-like object is required, not 'str'
```

beduffy committed:
Removed more comments, uneccessary code and prf_div default 16 commented

- TODO for not dealing with raw RF

beduffy committed:
Added README.md for install and run instructions

beduffy committed:
Made it work for python3 only

- b'<' and b'.' needed for byte string operations in py3
- no encode hex for python3

beduffy commented:
Hi, the latest commits now contains a README.md with those install instructions. We were mostly running python2 (necessary with ROS 1) over the months but python requires slightly different operations regarding dealing with byte strings but the latest code now runs only on python3 on my machine. We would like to make it work for both python 2 and 3 but not sure how much time this will take. 

Let me know if it doesn't work for you and if you have any other questions

Rafat1699 commented:
Hi,
I am recently using the slmx4 sensor for one of my research projects, we need some information. I have collected some data using MATLAB. It took data over a sugarbeet pile. 
1.	What is the Bin here in the graph. Does it correlate to any kind of distance?
2.	The amplitude in the graph what does it refers to? Is it the signal strength?
3.	Does this sensor respond to any data related to food item or a pile of food item rather than metal?
![allimg](https://github.com/SensorLogicInc/modules/assets/162401867/f1ac3c03-67a0-4867-aae7-b10185dd0973)
